File: mobile/androidApp/src/main/java/com/tinsu/mobile/ui/connection/ConnectionEditScreen.kt
package com.tinsu.mobile.ui.connection

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Error
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import com.tinsu.mobile.connection.ConnectionConfig
import com.tinsu.mobile.connection.ConnectionErrorType
import com.tinsu.mobile.connection.ConnectionListViewModel
import com.tinsu.mobile.connection.ConnectionRepository
import com.tinsu.mobile.connection.ConnectionTestState
import com.tinsu.mobile.connection.SessionInfo
import com.tinsu.mobile.connection.TransportType
import com.tinsu.mobile.ui.components.DestructiveButton
import com.tinsu.mobile.ui.components.PrimaryButton
import com.tinsu.mobile.ui.components.SecondaryButton
import com.tinsu.mobile.ui.theme.MonospaceCodeStyle
import com.tinsu.mobile.ui.theme.TinsuSpacing
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.koin.compose.koinInject

/**
 * Connection edit screen for creating and editing connections.
 *
 * @param existingConnection The connection to edit, or null for a new connection
 * @param viewModel The ConnectionListViewModel injected via Koin
 * @param onSave Callback when the connection is saved successfully
 * @param onCancel Callback when the user cancels editing
 * @param modifier Modifier for the screen
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ConnectionEditScreen(
    existingConnection: ConnectionConfig? = null,
    repository: ConnectionRepository = koinInject(),
    viewModel: ConnectionListViewModel = koinInject(),
    onSave: () -> Unit,
    onCancel: () -> Unit,
    modifier: Modifier = Modifier
) {
    val scrollBehavior = TopAppBarDefaults.pinnedScrollBehavior()
    val coroutineScope = rememberCoroutineScope()

    // Form state
    var displayName by remember { mutableStateOf(existingConnection?.displayName ?: "") }
    var host by remember { mutableStateOf(existingConnection?.host ?: "") }
    var port by remember { mutableStateOf(existingConnection?.port?.toString() ?: "22") }
    var username by remember { mutableStateOf(existingConnection?.username ?: "") }
    var transportType by remember { mutableStateOf(existingConnection?.transportType ?: TransportType.SSH) }
    var sshKeyAlias by remember { mutableStateOf(existingConnection?.sshKeyAlias ?: "") }

    // Validation errors
    var displayNameError by remember { mutableStateOf<String?>(null) }
    var hostError by remember { mutableStateOf<String?>(null) }
    var usernameError by remember { mutableStateOf<String?>(null) }
    var portError by remember { mutableStateOf<String?>(null) }
    var generalError by remember { mutableStateOf<String?>(null) }

    val isEditing = existingConnection != null

    // Test connection state
    val testStateMap by viewModel.testState.collectAsState()
    val stateKey = existingConnection?.id ?: "__new_${host}_${port}"
    val testState = testStateMap[stateKey] ?: ConnectionTestState.Idle
    val isTested = viewModel.isConnectionTested(stateKey)

    // For new connections, Save is disabled until test passes
    val canSave = isEditing || isTested

    fun validateAndSave() {
        displayNameError = null
        hostError = null
        usernameError = null
        portError = null
        generalError = null

        var isValid = true

        if (displayName.trim().isEmpty()) {
            displayNameError = "Display name is required"
            isValid = false
        } else if (displayName.trim().length > 50) {
            displayNameError = "Display name must be 50 characters or less"
            isValid = false
        }

        if (host.trim().isEmpty()) {
            hostError = "Host is required"
            isValid = false
        } else if (host.trim().length > 255) {
            hostError = "Host must be 255 characters or less"
            isValid = false
        }

        if (username.trim().isEmpty()) {
            usernameError = "Username is required"
            isValid = false
        } else if (username.trim().length > 100) {
            usernameError = "Username must be 100 characters or less"
            isValid = false
        }

        val portNum = port.toIntOrNull()
        if (portNum == null) {
            portError = "Port must be a number"
            isValid = false
        } else if (portNum < 1 || portNum > 65535) {
            portError = "Port must be between 1 and 65535"
            isValid = false
        }

        if (!isValid) return

        if (!isEditing && !isTested) {
            generalError = "Test your connection before saving"
            return
        }

        val portValue = portNum!!

        val config = existingConnection?.copy(
            displayName = displayName.trim(),
            host = host.trim(),
            port = portValue,
            username = username.trim(),
            transportType = transportType,
            sshKeyAlias = sshKeyAlias.trim().ifEmpty { null }
        ) ?: ConnectionConfig(
            displayName = displayName.trim(),
            host = host.trim(),
            port = portValue,
            username = username.trim(),
            transportType = transportType,
            sshKeyAlias = sshKeyAlias.trim().ifEmpty { null }
        )

        coroutineScope.launch {
            val result = if (isEditing) {
                repository.updateConnection(config)
            } else {
                repository.createConnection(config)
            }

            when (result) {
                is com.tinsu.mobile.util.Result.Success -> onSave()
                is com.tinsu.mobile.util.Result.Failure -> {
                    generalError = result.error.userMessage
                }
            }
        }
    }

    fun triggerTestConnection() {
        val portNum = port.toIntOrNull() ?: return
        val config = existingConnection?.copy(
            host = host.trim(),
            port = portNum,
            username = username.trim(),
            sshKeyAlias = sshKeyAlias.trim().ifEmpty { null }
        ) ?: ConnectionConfig(
            displayName = displayName.trim(),
            host = host.trim(),
            port = portNum,
            username = username.trim(),
            transportType = transportType,
            sshKeyAlias = sshKeyAlias.trim().ifEmpty { null }
        )
        viewModel.testConnection(config)
    }

    Scaffold(
        modifier = modifier.nestedScroll(scrollBehavior.nestedScrollConnection),
        topBar = {
            TopAppBar(
                title = { Text(if (isEditing) "Edit Connection" else "New Connection") },
                navigationIcon = {
                    IconButton(onClick = onCancel) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = "Cancel"
                        )
                    }
                },
                scrollBehavior = scrollBehavior
            )
        }
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .padding(TinsuSpacing.ContentMargin)
                .verticalScroll(rememberScrollState())
                .imePadding(),
            verticalArrangement = Arrangement.spacedBy(TinsuSpacing.ItemSpacing)
        ) {
            // General error message
            generalError?.let { error ->
                Text(
                    text = error,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.error
                )
            }

            // Display Name field
            OutlinedTextField(
                value = displayName,
                onValueChange = {
                    displayName = it
                    displayNameError = null
                },
                label = { Text("Display Name") },
                placeholder = { Text("My Server") },
                isError = displayNameError != null,
                supportingText = displayNameError?.let { { Text(it) } },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
                colors = OutlinedTextFieldDefaults.colors(
                    errorBorderColor = MaterialTheme.colorScheme.error,
                    errorSupportingTextColor = MaterialTheme.colorScheme.error
                )
            )

            // Host field
            OutlinedTextField(
                value = host,
                onValueChange = {
                    host = it
                    hostError = null
                },
                label = { Text("Host") },
                placeholder = { Text("example.com or 192.168.1.100") },
                isError = hostError != null,
                supportingText = hostError?.let { { Text(it) } },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
                colors = OutlinedTextFieldDefaults.colors(
                    errorBorderColor = MaterialTheme.colorScheme.error,
                    errorSupportingTextColor = MaterialTheme.colorScheme.error
                )
            )

            // Port and Username row
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(TinsuSpacing.ItemSpacing)
            ) {
                OutlinedTextField(
                    value = port,
                    onValueChange = {
                        port = it
                        portError = null
                    },
                    label = { Text("Port") },
                    isError = portError != null,
                    supportingText = portError?.let { { Text(it) } },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                    modifier = Modifier.weight(1f),
                    colors = OutlinedTextFieldDefaults.colors(
                        errorBorderColor = MaterialTheme.colorScheme.error,
                        errorSupportingTextColor = MaterialTheme.colorScheme.error
                    )
                )

                OutlinedTextField(
                    value = username,
                    onValueChange = {
                        username = it
                        usernameError = null
                    },
                    label = { Text("Username") },
                    placeholder = { Text("user") },
                    isError = usernameError != null,
                    supportingText = usernameError?.let { { Text(it) } },
                    singleLine = true,
                    modifier = Modifier.weight(2f),
                    colors = OutlinedTextFieldDefaults.colors(
                        errorBorderColor = MaterialTheme.colorScheme.error,
                        errorSupportingTextColor = MaterialTheme.colorScheme.error
                    )
                )
            }

            // Transport Type Selection
            Text(
                text = "Transport",
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Row(
                horizontalArrangement = Arrangement.spacedBy(TinsuSpacing.ItemSpacing)
            ) {
                TransportType.entries.forEach { type ->
                    Button(
                        onClick = { transportType = type },
                        colors = if (transportType == type) {
                            ButtonDefaults.buttonColors(
                                containerColor = MaterialTheme.colorScheme.primary
                            )
                        } else {
                            ButtonDefaults.outlinedButtonColors()
                        }
                    ) {
                        Text(type.displayName)
                    }
                }
            }

            // SSH Key Alias
            OutlinedTextField(
                value = sshKeyAlias,
                onValueChange = { sshKeyAlias = it },
                label = { Text("SSH Key Alias") },
                placeholder = { Text("my-ssh-key") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
                textStyle = MonospaceCodeStyle
            )

            // --- Test Connection Section ---
            Spacer(modifier = Modifier.size(TinsuSpacing.ItemSpacing))

            TestConnectionSection(
                testState = testState,
                onTestClick = { triggerTestConnection() },
                isTesting = testState is ConnectionTestState.Testing
            )

            // Mandatory test notice for new connections
            if (!isEditing && !isTested) {
                Text(
                    text = "Test your connection before saving",
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    fontFamily = FontFamily.Monospace
                )
            }

            Spacer(modifier = Modifier.size(TinsuSpacing.ItemSpacing))

            // Action buttons
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(TinsuSpacing.ItemSpacing)
            ) {
                SecondaryButton(
                    text = "Cancel",
                    onClick = onCancel,
                    modifier = Modifier.weight(1f)
                )
                PrimaryButton(
                    text = if (isEditing) "Save" else "Create",
                    onClick = { validateAndSave() },
                    modifier = Modifier.weight(1f),
                    enabled = canSave
                )
            }

            if (isEditing) {
                Spacer(modifier = Modifier.size(TinsuSpacing.ItemSpacing))
                DestructiveButton(
                    text = "Delete Connection",
                    onClick = { /* Handle delete - shown on edit only */ },
                    modifier = Modifier.fillMaxWidth()
                )
            }
        }
    }
}

@Composable
private fun TestConnectionSection(
    testState: ConnectionTestState,
    onTestClick: () -> Unit,
    isTesting: Boolean
) {
    Column(verticalArrangement = Arrangement.spacedBy(TinsuSpacing.ItemSpacing)) {
        // Test Connection button
        Button(
            onClick = onTestClick,
            enabled = !isTesting,
            modifier = Modifier.fillMaxWidth(),
            colors = ButtonDefaults.outlinedButtonColors(
                containerColor = Color.Transparent,
                disabledContainerColor = Color.Transparent
            ),
            shape = RoundedCornerShape(4.dp)
        ) {
            if (isTesting) {
                CircularProgressIndicator(
                    modifier = Modifier.size(18.dp),
                    strokeWidth = 2.dp,
                    color = MaterialTheme.colorScheme.primary
                )
                Spacer(modifier = Modifier.size(8.dp))
                Text("Testing...", style = MaterialTheme.typography.labelMedium)
            } else {
                Text("Test Connection", style = MaterialTheme.typography.labelMedium)
            }
        }

        // Result display
        AnimatedVisibility(
            visible = testState is ConnectionTestState.Success || testState is ConnectionTestState.Failure,
            enter = fadeIn() + slideInVertically(),
            exit = fadeOut() + slideOutVertically()
        ) {
            when (testState) {
                is ConnectionTestState.Success -> TestSuccessCard(testState.sessionInfo)
                is ConnectionTestState.Failure -> TestFailureCard(testState)
                else -> {}
            }
        }
    }
}

@Composable
private fun TestSuccessCard(sessionInfo: SessionInfo) {
    val successColor = Color(0xFF3FB950)

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(successColor.copy(alpha = 0.1f))
            .border(1.dp, successColor.copy(alpha = 0.3f), RoundedCornerShape(8.dp))
            .padding(TinsuSpacing.CardPadding)
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(
                imageVector = Icons.Filled.CheckCircle,
                contentDescription = "Connected",
                tint = successColor,
                modifier = Modifier.size(20.dp)
            )
            Spacer(modifier = Modifier.size(12.dp))
            Column {
                Text(
                    text = "Connected successfully",
                    style = MaterialTheme.typography.bodyMedium,
                    color = successColor
                )
                Text(
                    text = "${sessionInfo.authenticatedAs}@${sessionInfo.host}:${sessionInfo.port}",
                    style = MonospaceCodeStyle,
                    color = successColor.copy(alpha = 0.7f)
                )
                if (sessionInfo.serverVersion.isNotBlank()) {
                    Text(
                        text = sessionInfo.serverVersion,
                        style = MonospaceCodeStyle.copy(fontSize = MonospaceCodeStyle.fontSize * 0.85f),
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
        }
    }
}

@Composable
private fun TestFailureCard(failure: ConnectionTestState.Failure) {
    val errorColor = Color(0xFFF85149)

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(errorColor.copy(alpha = 0.08f))
            .border(1.dp, errorColor.copy(alpha = 0.3f), RoundedCornerShape(8.dp))
            .padding(TinsuSpacing.CardPadding)
    ) {
        Column {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    imageVector = Icons.Filled.Error,
                    contentDescription = "Connection failed",
                    tint = errorColor,
                    modifier = Modifier.size(20.dp)
                )
                Spacer(modifier = Modifier.size(12.dp))
                Text(
                    text = errorLabel(failure.errorType),
                    style = MaterialTheme.typography.bodyMedium,
                    color = errorColor
                )
            }

            Spacer(modifier = Modifier.size(8.dp))

            Text(
                text = failure.message,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurface
            )

            if (failure.hints.isNotEmpty()) {
                Spacer(modifier = Modifier.size(8.dp))
                Column {
                    Text(
                        text = "Troubleshooting:",
                        style = MaterialTheme.typography.labelMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                    failure.hints.forEach { hint ->
                        Row(modifier = Modifier.padding(start = 8.dp, top = 2.dp)) {
                            Text(
                                text = "  > ",
                                style = MonospaceCodeStyle.copy(fontSize = MonospaceCodeStyle.fontSize * 0.85f),
                                color = Color(0xFFD29922)
                            )
                            Text(
                                text = hint,
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }
                }
            }
        }
    }
}

private fun errorLabel(errorType: ConnectionErrorType): String = when (errorType) {
    ConnectionErrorType.HOST_UNREACHABLE -> "Host Unreachable"
    ConnectionErrorType.AUTH_FAILED -> "Authentication Failed"
    ConnectionErrorType.TIMEOUT -> "Connection Timed Out"
    ConnectionErrorType.PORT_BLOCKED -> "Port Blocked"
    ConnectionErrorType.KEY_NOT_FOUND -> "SSH Key Not Found"
    ConnectionErrorType.NETWORK_ERROR -> "Network Error"
    ConnectionErrorType.UNKNOWN -> "Connection Failed"
}

File: mobile/iosApp/iosApp/ConnectionEditScreen.swift
import SwiftUI
import Shared

/// SwiftUI view for creating/editing a connection matching Android design.
struct ConnectionEditScreen: View {
    let existingConnection: ConnectionConfig?
    let repository: ConnectionRepository
    let viewModel: ConnectionListViewModel
    let onSave: () -> Void

    @State private var displayName: String
    @State private var host: String
    @State private var port: String
    @State private var username: String
    @State private var transportType: TransportType
    @State private var sshKeyAlias: String

    @State private var displayNameError: String?
    @State private var hostError: String?
    @State private var usernameError: String?
    @State private var portError: String?
    @State private var generalError: String?

    @State private var testState: ConnectionTestState = .idle
    @State private var isTested: Bool = false

    @Environment(\.dismiss) private var dismiss

    var isEditing: Bool { existingConnection != nil }

    /// State key matching the ViewModel's key logic
    private var stateKey: String {
        existingConnection?.id ?? "__new_\(host)_\(port)"
    }

    var canSave: Bool { isEditing || isTested }

    init(
        existingConnection: ConnectionConfig?,
        repository: ConnectionRepository,
        viewModel: ConnectionListViewModel,
        onSave: @escaping () -> Void
    ) {
        self.existingConnection = existingConnection
        self.repository = repository
        self.viewModel = viewModel
        self.onSave = onSave

        _displayName = State(initialValue: existingConnection?.displayName ?? "")
        _host = State(initialValue: existingConnection?.host ?? "")
        _port = State(initialValue: existingConnection?.port.description ?? "22")
        _username = State(initialValue: existingConnection?.username ?? "")
        _transportType = State(initialValue: existingConnection?.transportType ?? .ssh)
        _sshKeyAlias = State(initialValue: existingConnection?.sshKeyAlias ?? "")
    }

    var body: some View {
        Form {
            // General error message
            if let error = generalError {
                Section {
                    Text(error)
                        .font(TinsuTypography.body)
                        .foregroundColor(TinsuColors.error)
                }
            }

            // Display Name
            Section {
                TextField("Display Name", text: $displayName)
                    .onChange(of: displayName) { _ in
                        displayNameError = nil
                    }
                if let error = displayNameError {
                    Text(error)
                        .font(TinsuTypography.label)
                        .foregroundColor(TinsuColors.error)
                }
            } header: {
                Text("Connection Name")
            }

            // Host
            Section {
                TextField("Host", text: $host)
                    .keyboardType(.URL)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .onChange(of: host) { _ in
                        hostError = nil
                    }
                if let error = hostError {
                    Text(error)
                        .font(TinsuTypography.label)
                        .foregroundColor(TinsuColors.error)
                }
            } header: {
                Text("Server Address")
            } footer: {
                Text("Enter hostname (e.g., example.com) or IP address")
                    .font(TinsuTypography.label)
            }

            // Port and Username
            Section {
                HStack {
                    Text("Port")
                    Spacer()
                    TextField("22", text: $port)
                        .keyboardType(.numberPad)
                        .multilineTextAlignment(.trailing)
                        .onChange(of: port) { _ in
                            portError = nil
                        }
                }
                if let error = portError {
                    Text(error)
                        .font(TinsuTypography.label)
                        .foregroundColor(TinsuColors.error)
                }

                TextField("Username", text: $username)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .onChange(of: username) { _ in
                        usernameError = nil
                    }
                if let error = usernameError {
                    Text(error)
                        .font(TinsuTypography.label)
                        .foregroundColor(TinsuColors.error)
                }
            }

            // Transport Type
            Section {
                Picker("Transport", selection: $transportType) {
                    Text("SSH").tag(TransportType.ssh)
                    Text("mosh").tag(TransportType.mosh)
                }
                .pickerStyle(.segmented)
            } header: {
                Text("Connection Protocol")
            } footer: {
                Text("SSH is standard; mosh provides better resilience on unstable connections")
                    .font(TinsuTypography.label)
            }

            // SSH Key Alias
            Section {
                TextField("SSH Key Alias", text: $sshKeyAlias)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
            } header: {
                Text("SSH Key")
            } footer: {
                Text("Select or generate an SSH key before testing")
                    .font(TinsuTypography.label)
            }

            // --- Test Connection ---
            Section {
                // Test button
                Button(action: triggerTestConnection) {
                    HStack {
                        Spacer()
                        if case .testing = testState {
                            ProgressView()
                                .progressViewStyle(CircularProgressViewStyle(tint: TinsuColors.primary))
                            Text("Testing...")
                                .font(TinsuTypography.label)
                        } else {
                            Text("Test Connection")
                                .font(TinsuTypography.label)
                        }
                        Spacer()
                    }
                }
                .secondaryButtonStyle()
                .disabled(isTestStateTesting)

                // Result display
                switch testState {
                case .success(let sessionInfo):
                    TestSuccessCard(sessionInfo: sessionInfo)
                        .transition(.opacity.combined(with: .move(edge: .top)))
                case .failure(let errorType, let message, let hints):
                    TestFailureCard(
                        errorType: errorType,
                        message: message,
                        hints: hints
                    )
                    .transition(.opacity.combined(with: .move(edge: .top)))
                default:
                    EmptyView()
                }

                // Mandatory test notice for new connections
                if !isEditing && !isTested {
                    Text("Test your connection before saving")
                        .font(TinsuTypography.code)
                        .foregroundColor(TinsuColors.onSurfaceVariant)
                }
            } header: {
                Text("Connection Test")
            }

            // Actions
            Section {
                Button(action: validateAndSave) {
                    HStack {
                        Spacer()
                        Text(isEditing ? "Save Changes" : "Create Connection")
                            .frame(maxWidth: .infinity)
                        Spacer()
                    }
                }
                .primaryButtonStyle()
                .disabled(!canSave)

                Button("Cancel") {
                    dismiss()
                }
                .secondaryButtonStyle()
            }
        }
        .navigationTitle(isEditing ? "Edit Connection" : "New Connection")
        .navigationBarTitleDisplayMode(.inline)
    }

    private var isTestStateTesting: Bool {
        if case .testing = testState { return true }
        return false
    }

    // MARK: - Test Connection

    private func triggerTestConnection() {
        testState = .testing

        let portNum = Int(port) ?? 22
        let config = ConnectionConfig(
            id: existingConnection?.id,
            displayName: displayName.trimmingCharacters(in: .whitespaces),
            host: host.trimmingCharacters(in: .whitespaces),
            port: Int32(portNum),
            username: username.trimmingCharacters(in: .whitespaces),
            transportType: transportType,
            sshKeyAlias: sshKeyAlias.trimmingCharacters(in: .whitespaces).isEmpty ? nil : sshKeyAlias.trimmingCharacters(in: .whitespaces),
            sortOrder: existingConnection?.sortOrder ?? 0,
            lastConnectedAt: existingConnection?.lastConnectedAt,
            createdAt: existingConnection?.createdAt ?? Int64(Date().timeIntervalSince1970 * 1000),
            updatedAt: Int64(Date().timeIntervalSince1970 * 1000)
        )

        viewModel.testConnection(config: config)

        // Observe result from ViewModel flow
        Task { @MainActor in
            // Small delay to allow the ViewModel to process
            try? await Task.sleep(nanoseconds: 500_000_000)
            let state = viewModel.getTestState(connectionId: stateKey)
            switch state {
            case let s as Shared.ConnectionTestState.Success:
                testState = .success(sessionInfo: s.sessionInfo)
                isTested = true
            case let f as Shared.ConnectionTestState.Failure:
                testState = .failure(errorType: f.errorType, message: f.message, hints: f.hints)
            default:
                break
            }
        }
    }

    // MARK: - Validation

    private func validateAndSave() {
        displayNameError = nil
        hostError = nil
        usernameError = nil
        portError = nil
        generalError = nil

        var isValid = true

        let trimmedDisplayName = displayName.trimmingCharacters(in: .whitespaces)
        if trimmedDisplayName.isEmpty {
            displayNameError = "Display name is required"
            isValid = false
        } else if trimmedDisplayName.count > 50 {
            displayNameError = "Display name must be 50 characters or less"
            isValid = false
        }

        let trimmedHost = host.trimmingCharacters(in: .whitespaces)
        if trimmedHost.isEmpty {
            hostError = "Host is required"
            isValid = false
        } else if trimmedHost.count > 255 {
            hostError = "Host must be 255 characters or less"
            isValid = false
        }

        let trimmedUsername = username.trimmingCharacters(in: .whitespaces)
        if trimmedUsername.isEmpty {
            usernameError = "Username is required"
            isValid = false
        } else if trimmedUsername.count > 100 {
            usernameError = "Username must be 100 characters or less"
            isValid = false
        }

        guard let portNum = Int(port), portNum >= 1, portNum <= 65535 else {
            portError = "Port must be between 1 and 65535"
            isValid = false
            return
        }

        guard isValid else { return }

        if !isEditing && !isTested {
            generalError = "Test your connection before saving"
            return
        }

        let config = ConnectionConfig(
            id: existingConnection?.id,
            displayName: trimmedDisplayName,
            host: trimmedHost,
            port: Int32(portNum),
            username: trimmedUsername,
            transportType: transportType,
            sshKeyAlias: sshKeyAlias.trimmingCharacters(in: .whitespaces).isEmpty ? nil : sshKeyAlias.trimmingCharacters(in: .whitespaces),
            sortOrder: existingConnection?.sortOrder ?? 0,
            lastConnectedAt: existingConnection?.lastConnectedAt,
            createdAt: existingConnection?.createdAt ?? Int64(Date().timeIntervalSince1970 * 1000),
            updatedAt: Int64(Date().timeIntervalSince1970 * 1000)
        )

        if isEditing {
            repository.updateConnection(config: config) { _, error in
                DispatchQueue.main.async {
                    if let error = error {
                        self.generalError = error.localizedDescription
                    } else {
                        self.onSave()
                        self.dismiss()
                    }
                }
            }
        } else {
            repository.createConnection(config: config) { _, error in
                DispatchQueue.main.async {
                    if let error = error {
                        self.generalError = error.localizedDescription
                    } else {
                        self.onSave()
                        self.dismiss()
                    }
                }
            }
        }
    }
}

// MARK: - Test Result Cards

private struct TestSuccessCard: View {
    let sessionInfo: SessionInfo

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 8) {
                Image(systemName: "checkmark.circle.fill")
                    .foregroundColor(TinsuColors.success)
                    .font(.system(size: 18))
                Text("Connected successfully")
                    .font(TinsuTypography.body)
                    .foregroundColor(TinsuColors.success)
            }
            Text("\(sessionInfo.authenticatedAs)@\(sessionInfo.host):\(sessionInfo.port)")
                .font(TinsuTypography.code)
                .foregroundColor(TinsuColors.success.opacity(0.7))
            if !sessionInfo.serverVersion.isEmpty {
                Text(sessionInfo.serverVersion)
                    .font(TinsuTypography.lineNumbers)
                    .foregroundColor(TinsuColors.onSurfaceVariant)
            }
        }
        .padding(TinsuSpacing.cardPadding)
        .background(TinsuColors.success.opacity(0.1))
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(TinsuColors.success.opacity(0.3), lineWidth: 1)
        )
        .cornerRadius(8)
    }
}

private struct TestFailureCard: View {
    let errorType: ConnectionErrorType
    let message: String
    let hints: [String]

    private var errorLabel: String {
        switch errorType {
        case .hostUnreachable: return "Host Unreachable"
        case .authFailed: return "Authentication Failed"
        case .timeout: return "Connection Timed Out"
        case .portBlocked: return "Port Blocked"
        case .keyNotFound: return "SSH Key Not Found"
        case .networkError: return "Network Error"
        default: return "Connection Failed"
        }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 8) {
                Image(systemName: "xmark.circle.fill")
                    .foregroundColor(TinsuColors.error)
                    .font(.system(size: 18))
                Text(errorLabel)
                    .font(TinsuTypography.body)
                    .foregroundColor(TinsuColors.error)
            }

            Text(message)
                .font(TinsuTypography.body)
                .foregroundColor(TinsuColors.onBackground)

            if !hints.isEmpty {
                VStack(alignment: .leading, spacing: 4) {
                    Text("Troubleshooting:")
                        .font(TinsuTypography.label)
                        .foregroundColor(TinsuColors.onSurfaceVariant)

                    ForEach(hints, id: \.self) { hint in
                        HStack(alignment: .top, spacing: 4) {
                            Text(">")
                                .font(TinsuTypography.code)
                                .foregroundColor(TinsuColors.warning)
                            Text(hint)
                                .font(TinsuTypography.label)
                                .foregroundColor(TinsuColors.onSurfaceVariant)
                        }
                        .padding(.leading, 8)
                    }
                }
            }
        }
        .padding(TinsuSpacing.cardPadding)
        .background(TinsuColors.error.opacity(0.08))
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(TinsuColors.error.opacity(0.3), lineWidth: 1)
        )
        .cornerRadius(8)
    }
}

// MARK: - Local Test State (mirrors KMP sealed class for SwiftUI)

private enum ConnectionTestState {
    case idle
    case testing
    case success(sessionInfo: SessionInfo)
    case failure(errorType: ConnectionErrorType, message: String, hints: [String])
}

// MARK: - Preview

#Preview("New Connection") {
    NavigationView {
        ConnectionEditScreen(
            existingConnection: nil,
            repository: FakeConnectionRepository(),
            viewModel: ConnectionListViewModel(repository: FakeConnectionRepository()),
            onSave: {}
        )
    }
    .preferredColorScheme(.dark)
}

#Preview("Edit Connection") {
    NavigationView {
        ConnectionEditScreen(
            existingConnection: ConnectionConfig(
                id: "1",
                displayName: "My Server",
                host: "example.com",
                port: 22,
                username: "user",
                transportType: .ssh,
                sshKeyAlias: "my-key",
                sortOrder: 0,
                lastConnectedAt: KotlinLong(value: Int64(Date().timeIntervalSince1970 * 1000)),
                createdAt: Int64(Date().timeIntervalSince1970 * 1000),
                updatedAt: Int64(Date().timeIntervalSince1970 * 1000)
            ),
            repository: FakeConnectionRepository(),
            viewModel: ConnectionListViewModel(repository: FakeConnectionRepository()),
            onSave: {}
        )
    }
    .preferredColorScheme(.dark)
}
